src/host/getset.cpp

```diff
@@ -437,19 +441,29 @@ void ApiRoutines::GetNumberOfConsoleMouseButtonsImpl(ULONG& buttons) noexcept
             screenInfo.GetStateMachine().ResetState();
         }
 
-        // if we changed rendering modes then redraw the output buffer,
-        // but only do this if we're not in conpty mode.
-        if (!gci.IsInVtIoMode() &&
-            (WI_IsFlagSet(dwNewMode, ENABLE_VIRTUAL_TERMINAL_PROCESSING) != WI_IsFlagSet(dwOldMode, ENABLE_VIRTUAL_TERMINAL_PROCESSING) ||
-             WI_IsFlagSet(dwNewMode, ENABLE_LVB_GRID_WORLDWIDE) != WI_IsFlagSet(dwOldMode, ENABLE_LVB_GRID_WORLDWIDE)))
+        // if we changed rendering modes then redraw the output buffer.
+        if (WI_IsAnyFlagSet(diff, ENABLE_VIRTUAL_TERMINAL_PROCESSING | ENABLE_LVB_GRID_WORLDWIDE))
         {
-            auto* pRender = ServiceLocator::LocateGlobals().pRender;
-            if (pRender)
+            if (const auto pRender = ServiceLocator::LocateGlobals().pRender)
             {
                 pRender->TriggerRedrawAll();
             }
         }
 
+        auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+        if (const auto io = gci.GetVtIoForBuffer(&context))
+        {
+            if (WI_IsFlagSet(diff, ENABLE_WRAP_AT_EOL_OUTPUT))
+            {
+                io->WriteDECAWM(WI_IsFlagSet(dwNewMode, ENABLE_WRAP_AT_EOL_OUTPUT));
+            }

```

Comment:
Previously, `WriteCharsLegacy` used a small local buffer to accumulate plain text first (to translate control chars, etc.). That buffer size effectively determined how many columns the cursor could move at a time. After every fast pass we would call `AdjustCursorPosition` which would handle `ENABLE_WRAP_AT_EOL_OUTPUT = false`.

I haven't tested DOS, but between Windows XP and Windows 11 21H2 the buffer size used to be 100 characters. Assuming no wide chars, if the window was smaller than 100 columns, then every fast-pass would cover an entire row of text, which would mean that the cursor gets reset to column 0 after every write. If the window was wider than that, the result was pretty much random, depending on the window size. To me this has every indication of an unintentional bug.

What's really funny is that I found a comment in the Windows XP source code which appears to complain about `ENABLE_WRAP_AT_EOL_OUTPUT = false` not working like it used to, because it wouldn't leave the cursor at the end of the row.

Additionally, the MSDN docs stated all the time between Windows XP up to today:
> **ENABLE_WRAP_AT_EOL_OUTPUT**
> [...] If this mode is disabled, the last character in the row is overwritten with any subsequent characters.

Wine also implements it that way (as documented by MSDN).

Given the discrepancy between the things listed above, I guess I should stick with how it actually used to work in conhost. However, that would mean we'd have to restore the 100-char buffer in `WriteCharsLegacy` and keep that one forever too. Hmm... Maybe I'll stick with `DECAWM` for now and fix this in a future PR.

Reply:
FYI I updated the PR description with a list of quirks/issues. I'll collapse some of these comments, so I can better track the remaining things that I need to fix.